File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/structural/decorator/LoggerDecorator.java
package design.pattern.java.patterns.structural.decorator;

public class LoggerDecorator implements ServiceInterface {
    private final ServiceInterface service;

    public LoggerDecorator(ServiceInterface service) {
        this.service = service;
    }

    @Override
    public String run(String value) {
        this.log(value);
        return service.run(value);
    }

    protected void log(String value) {
        System.out.println("Logging: ".concat(value));
    }

}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/structural/decorator/LoggerDecorator.kt
package design.pattern.kotlin.patterns.structural.decorator

open class LoggerDecorator(private val service: ServiceInterface) : ServiceInterface {
    override fun run(value: String): String {
        log(value)
        return service.run(value)
    }

    internal open fun log(value: String) {
        println("Logging: $value")
    }
}